### PR TITLE
[RND-298] Common collection schema issue

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/Utility.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ export function isDescriptor(property: EntityProperty): boolean {
 
 export function uncapitalize(text: string): string {
   if (text == null || typeof text !== 'string') return '';
+  // Handle text like "URI" -> "uri"
+  if (text === text.toUpperCase()) return text.toLowerCase();
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/Utility.ts:28]: URI was turning into uRI, should be uri

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts [hunk 2]
```diff
@@ -37,18 +39,21 @@ function parentPrefixRemovalConvention(property: EntityProperty): string {
...
  */
-function apiFullName(property: EntityProperty, { removeCollectionPrefixes }): string {
-  if (property.isCollection && removeCollectionPrefixes) {
+function apiFullName(property: EntityProperty, { removePrefixes }: NamingOptions): string {
+  if (property.isCollection && removePrefixes) {
     return uncapitalize(pluralize(parentPrefixRemovalConvention(property)));
   }
-  if (property.isCollection && !removeCollectionPrefixes) {
+  if (property.type === 'common' && removePrefixes) {
+    return uncapitalize(parentPrefixRemovalConvention(property));
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts:54]: Needed to remove prefixes on all commons e.g. Assessment.AssessmentPeriod -> "period" not "assessmentPeriod"

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts [hunk 3]
```diff
@@ -37,18 +39,21 @@ function parentPrefixRemovalConvention(property: EntityProperty): string {
  * API descriptor reference property names are suffixed with "Descriptor"
  */
 function apiDescriptorReferenceName(property): string {
-  return normalizeDescriptorSuffix(uncapitalize(property.fullPropertyName));
+  return normalizeDescriptorSuffix(uncapitalize(property.metaEdName));
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts:42]: No rolename prefix on descriptor references e.g. Assessment.AssessedGradeLevelDescriptor -> gradeLevelDescriptor

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts [hunk 4]
```diff
@@ -69,47 +74,54 @@ function apiReferenceName(property): string {
...
-    decollisionedTopLevelName: apiTopLevelName(property, { removeCollectionPrefixes: false }),
-    fullName: apiFullName(property, { removeCollectionPrefixes: true }),
+    topLevelName: apiTopLevelName(property, { removePrefixes: true }),
+    decollisionedTopLevelName: apiTopLevelName(property, { removePrefixes: false }),
+    fullName: apiFullName(property, { removePrefixes: true }),
     isReferenceCollection,
     referenceCollectionName: isReferenceCollection ? apiReferenceName(property) : '',
     isDescriptorCollection,
     descriptorCollectionName: isDescriptorCollection ? apiDescriptorReferenceName(property) : '',
+    isCommonCollection,
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts:124]: Added common collection flag like for reference and descriptor collections

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 5]
```diff
@@ -3,6 +3,8 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
 
+/* eslint-disable no-use-before-define */
+
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:7]: Recursion means use-before-define happens multiple times

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 6]
```diff
@@ -29,6 +31,18 @@ const descriptorSchema = Joi.object({
   description: Joi.string().optional(),
 });
 
+/**
+ * Returns a new PropertyModifier that is the concatenation of two. Used for Commons and sub-Commons,
+ * where there is a chain of parent modifiers that cannot be completely pre-computed
+ * (without a different design, like pre-computing all possible paths).
+ */
+function propertyModifierConcat(p1: PropertyModifier, p2: PropertyModifier): PropertyModifier {
+  return {
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:40]: PropertyModifier tracks property naming modifications that occur as you go from parent entity -> property. This is to bridge the modifiers when there are multiple entities involved e.g. entity -> common property -> common entity -> other property

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 7]
```diff
@@ -49,23 +62,50 @@ function joiTypeForReferentialProperty(property: ReferentialProperty, propertyMo
 }
 
 /**
- * Returns a Joi schema fragment that specifies the API body element shape
- * corresponding to the given common property.
+ * Returns a Joi schema of the properties of a referenced Common entity, handling issues of parent naming,
+ * and choice/inline common property pull-up
  */
-function joiTypeForCommonProperty(property: CommonProperty, propertyModifier: PropertyModifier) {
+function schemaFromPropertiesOfCommon(property: CommonProperty, propertyModifier: PropertyModifier) {
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:68]: This block of three functions re-does common property handling, mostly adding common collection support.

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/PropertyCollectingEnhancer.ts [hunk 8]
```diff
@@ -13,7 +13,7 @@ import { collectProperties } from './BasePropertyCollectingEnhancer';
  * Accumulates properties that belong under an entity in the API body.
  */
 export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
-  getAllEntitiesOfType(metaEd, 'domainEntity', 'association').forEach((entity) => {
+  getAllEntitiesOfType(metaEd, 'domainEntity', 'association', 'common').forEach((entity) => {
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/PropertyCollectingEnhancer.ts:16]: Start getting PropertyModifier information on properties of Commons

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/test/enhancer/JoiTestHelper.ts [hunk 9]
```diff
@@ -37,3 +37,14 @@ export function expectSubschemaScalarArray(joiSchema: any, schemaProperty: Schem
   expect(subschema.schema.type).toBe(schemaProperty.type);
   return subschema.schema;
 }
+
+export function expectSubschemaArray(joiSchema: any, schemaProperties: SchemaProperty[]) {
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/test/enhancer/JoiTestHelper.ts:41]: Needed more generalized inspection of array schemas for testing

File: Meadowlark-js/tests/http/local.common.collection.issue.http [hunk 10]
```diff
@@ -0,0 +1,168 @@
+#### These are the manual components of the automated system tests
```
Comment (bradbanister) [inline Meadowlark-js/tests/http/local.common.collection.issue.http:1]: Full complex resource documents to use to test schema. If they work, they'll fail on reference validation - after schema validation.

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 11]
```diff
@@ -3,6 +3,8 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
 
+/* eslint-disable no-use-before-define */
+
 // TODO: Need to upgrade to latest joi (https://joi.dev/), which is now top-level and no longer under hapi
```
Comment (stephenfuqua) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:8]: I think you said we would want to _replace_ Joi at some point. Can you create a ticket for that and add the ticket number as a reference in the line above?

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts (hunk 3 above)
Comment (stephenfuqua) [reply Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts:42]: Another surprising error (`local.33b.http` line 466): posting to `studentSchoolAssociations`:
```json
{
  "message": [
    "gradeLevelDescriptor is required",
    "entryGradeLevelDescriptor is not allowed"
  ]
}
```

Comment (bradbanister) [reply Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiPropertyMappingEnhancer.ts]: Fixed - unit test added

**Pull request merged**